resolver/resolver.go

```diff
@@ -0,0 +1,84 @@
+package resolver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync/atomic"
+)
+
+type resolver struct {
+	addresses []string
+	dialer    *net.Dialer
+	idx       uint64
+}
+
+// NewResolver - create a new instance of a dns resolver for plugging
+// into net.DefaultResolver.  Addresses should be a list of
+// ip addresses and optional port numbers, separated by colon.
+// For example: 1.2.3.4:53 and 1.2.3.4 are both valid.  In the absence
+// of a port number, 53 will be used instead.
+func NewResolver(addresses []string) (*net.Resolver, error) {
+	normalAddresses, err := normalizeResolverAddresses(addresses)
+	if err != nil {
+		return nil, err
+	}
+	return &net.Resolver{
+		PreferGo: true,
+		Dial:     (&resolver{addresses: normalAddresses, dialer: &net.Dialer{}}).dial,
+	}, nil
+}
+
+func normalizeResolverAddresses(addresses []string) ([]string, error) {
+	if len(addresses) == 0 {
+		return nil, errors.New("must specify at least resolver address")
+	}
+	normalAddresses := make([]string, len(addresses))
+	for i, addr := range addresses {
```

Comment:
You can re-use https://golang.org/pkg/net/#SplitHostPort for this purpose.

Reply:
That gives an error unless both host and port are present.  My version allows for just host or combination of host:port

Reply:
Yeah, but you can ignore the error. The host and port are still returned.

Reply:
No, that's not how it works at all.

https://play.golang.org/p/0rTnN5eoAj2

Reply:
```go
for i, addr := range addrs {
    host, port, _ := net.SplitHostPort(addr)
    if port == "" {
        port = "53"
    }
    normal[i] = net.JoinHostPort(host, port)
}
```

Reply:
Woops, never mind. I was quite convinced that was the case but it isn't.

Reply:
if there is no colon in the hostport string passed to SplitHostPort, host will be blank, port will be blank, error is the only part of the return with non-zero value.

Reply:
Then can we ensure the addr has a port and then reuse `net.SplitHostPort`?

Reply:
sure

Reply:
I meant `net.JoinHostPort`, instead of `fmt.Sprintf`. So, last requests here:

1. Reuse `net.JoinHostPort`
2. Rename `pu16` to `pn` (port number)
3. Rename `normalAddresses` to `normal`

Reply:
Actually it just occurred to me, my naive implementation breaks on parsing ipv6.  So far all of our labs are ipv4

Reply:
@nathanejohnson: Ah, well spotted!